[improve][doc] Release policy and matrix (#433)
apache/pulsar-site: contribute/release-policy.md

---
id: release-policy
title: Release policy
---

> ## Release semantics

Reviewer: Should it be semantic version?

Author: I lower the relation to semver. Please see the latest commit f29a51cc9178f8a02acf5eb9d99d56e1c19abb1a

The Pulsar project follows a variant of [Semantic Versioning](http://semver.org/spec/v2.0.0.html). Existing releases can expect patches for bugs and security vulnerabilities. New features will target minor releases. The difference is that a major version bump will not carry any special meaning in terms of "big features" included in the release or breaking API changes. Instead, it would simply signal a new long-term support (LTS) release.

For example,

* 2.10.0 is a feature release;
* 2.10.1 is a patch release;
* 2.11.0 is a feature release;
* 3.0.0 is the first LTS release;
* 3.1.0 is a feature release;
* 3.2.0 is a feature release;

> * 3.2.1 is a patch release;

Reviewer 2: Change to "major/minor" to keep consistent? Can you check all the occurencies in the whole doc?

Author: Please take a look at:

* https://github.com/apache/pulsar-site/pull/433#discussion_r1115172798
* https://github.com/apache/pulsar-site/pull/433#issuecomment-1441130956

I can update all the occurrences but we need some consensus first.

Reviewer: We should also add 

`3.0.1 is the a patch release of the LTS version`

Author: SGTM.

| Patch release   | When it is ready  | N/A            | N/A              |

This can be translated into:

* The last 2 LTS releases and the last 2 feature releases are supported.
* Security patches are provided for the past 3 LTS releases and 2 feature releases

Therefore, users can choose between stay in an LTS release until they are ready to jump into the next LTS, or try the latest releases which contains required features.

## Supported Versions

````mdx-code-block
import SupportedVersionsTable from "@site/src/components/SupportedVersionsTable";

<SupportedVersionsTable />
````

## Roadmap for release plans

The next release of Pulsar is 3.0.0, and it has the planned timeline as:

* 2023-04-11 - RC-1
* 2023-04-18 - RC-2
* 2023-04-25 - RC-3
* 2023-05-02 - Announce 3.0 Release

## Release cycles

Generally, one committer shall volunteer as the release manager (RM) for a specific release.

For feature releases and LTS releases, the last 3 weeks of the release cycle will be marked as a code-freeze period. The RM will branch off from master, and the RM is also responsible for selecting the changes that will be cherry-picked in the release branch.

> 

Reviewer: Move them

```
* 2023-04-11 - RC-1
 * 2023-04-18 - RC-2
 * 2023-04-25 - RC-3
 * 2023-05-02 - Announce 3.0 Release
```

here looks better.

Author: I'll read it as remove the RoadMap part but use 3.0.0 as an example for the release cycle section.

Author: Updated.

---
id: release-policy
title: Release policy
---

## Release semantics

> The Pulsar project follows a variant of Semantic Versioning (semver), which mapping `z.minor.patch` to `LTS.feature.patch`.

Author: ```suggestion
The Pulsar project follows a variant of Semantic Versioning (semver), which replacing `major.minor.patch` with `LTS.feature.patch`.
```

Concretely, existing releases can expect patches for bugs and security vulnerabilities. New features will target to feature releases. A "major" version bump will not carry any special meaning in terms of "big features" included in the release or breaking API changes. Instead, it would simply signal a new long-term support (LTS) release.

For example,

* 2.10.0 is a feature release;
* 2.10.1 is a patch release;
* 2.11.0 is a feature release;
* 3.0.0 is the first LTS release;
* 3.0.1 is a patch release of the LTS release;
* 3.1.0 is a feature release;
* 3.2.0 is a feature release;
* 3.2.1 is a patch release;

> * 4.0.0 is a LTS release.

Reviewer 3: ```suggestion
* 4.0.0 is an LTS release.
```

* 3.2 -> 4.4 -> 3.2 is OK;
* 3.2 -> 5.0 is _not_ OK.

## Release frequency and support expectation

|                 | Release frequency | Active Support | Security Support |
|-----------------|-------------------|----------------|------------------|
| LTS release     | Every 18 months   | 24 months      | 36 months        |
| Feature release | Every 3 months    | 6 months       | 6 months         |
| Patch release   | When it is ready  | N/A            | N/A              |

This can be translated into:

* The last 2 LTS releases and the last 2 feature releases are supported.
* Security patches are provided for the past 3 LTS releases and 2 feature releases

Therefore, users can choose between stay in an LTS release until they are ready to jump into the next LTS, or try the latest releases which contains required features.

## Supported Versions

````mdx-code-block
import SupportedVersionsTable from "@site/src/components/SupportedVersionsTable";

<SupportedVersionsTable />
````

## Roadmap for release plans

The next release of Pulsar is 3.0.0, and it has the planned timeline as:

* 2023-04-11 - RC-1
* 2023-04-18 - RC-2
* 2023-04-25 - RC-3
* 2023-05-02 - Announce 3.0 Release

> ## Release cycles

Reviewer 2: Suggest re-organize this part to "steps" (which can be based on "time" or "task category" and add a flowchart to indicate the process more clearly.

Author: Fixed by 3613ed52d6835479472a7281b70e949fc06f58f7.

A flowchart may be made later.

| LTS release     | Every 18 months   | 24 months      | 36 months        |
| Feature release | Every 3 months    | 6 months       | 6 months         |
| Patch release   | When it is ready  | N/A            | N/A              |

This can be translated into:

* The last 2 LTS releases and the last 2 feature releases are supported.
* Security patches are provided for the past 3 LTS releases and 2 feature releases

Therefore, users can choose between stay in an LTS release until they are ready to jump into the next LTS, or try the latest releases which contains required features.

## Supported Versions

| Version | Released      | Active Support	 | Security Support | Latest |
|---------|---------------|-----------------|------------------|--------|
| 2.10    | April 2022    | April 2023      | April 2023       | 2.10.3 |
| 2.9     | November 2021 | November 2022   | November 2022    | 2.9.4  |
| 2.8     | June 2021     | June 2022       | June 2022        | 2.8.4  |
| 2.7     | November 2020 | November 2021   | November 2021    | 2.7.5  |
| 2.6     | June 2020     | June 2021       | June 2021        | 2.6.4  |
| 2.5     | January 2020  | January 2021    | January 2021     | 2.5.2  |

> | 2.4     | July 2019     | July 2020       | July 2020        | 2.4.2  |

Author: This table is possibly to be data-driven, perhaps by `data/release-pulsar.js`.

I'm investigating for doing it programatically, but leave it as is for the first review so that we know what it basically looks like.

The final result should be as https://endoflife.date/python where:

1. Distinguish supported/unsupported versions by color.
2. Calculate the time interval.
3. Link to release notes.
4. Del line for unsupported versions.

Author: Finished.

## Roadmap for release plans

The next release of Pulsar is 3.0.0, and it has the planned timeline as:

* 2023-04-11 - RC-1
* 2023-04-18 - RC-2
* 2023-04-25 - RC-3

> * 2023-05-02 - Announce 3.0 Release

Author: I hesitate about publishing these TODOs since it should be continuously updated manually and it's quite an ideal case - I'm afraid that we will almost break the "promised" timeline.

Reviewer 2: ➕1 on showing the TODOs because it's a way to demonstrate the community is thinking about the future directions and has a plan on it. The positive influence is more than negative effect (fail to hit the DDL).

Reviewer: I think we don't need to add RC-1, RC-2 here. It only for the internal release process. The documentation will be available for all users. So it's better to keep it simple, the next release 3.0.0 will available in 2023-05-02

---
id: release-policy
title: Release policy
---

## Release semantics

> The Pulsar project follows a variant of [Semantic Versioning](http://semver.org/spec/v2.0.0.html). Existing releases can expect patches for bugs and security vulnerabilities. New features will target minor releases. The difference is that a major version bump will not carry any special meaning in terms of "big features" included in the release or breaking API changes. Instead, it would simply signal a new long-term support (LTS) release.

Reviewer 3: `New features will target minor releases.`  ---- I guess it refers to "major", like 2.10?

Author: When we use `marjor.minor.patch` concept, it's the minor.

Reviewer: We don't strictly follow the semver spec. We should not introduce incompatible APIs from 3.0.0 to 4.0.0

The LTS version is for stability since we will have many feature and patch releases before bumping a new LTS version. IMO, we should use LTS, feature, and patch releases in the feature. The major release and minor release is confusing.

And the patch release can also happen on a LTS version, like 3.0.1, 3.0.2.

Author: Sounds reasonable. Let me try to reword it.

For example,

* 2.10.0 is a feature release;

> * 2.10.1 is a patch release;

Reviewer 3: We need to align on the wording and consistently use it throughout the doc site. 
@codelipenghui and @BewareMyPower mentioned there was no `patch release` term for Pulsar, so we named it `minor` in contrast to `major` in the release process doc. If we decide to go with `feature/patch`, better eliminate `major/minor` to avoid confusing and guessing.

Reviewer: Yes, before, we only used major release and minor release. Actually, major release == feature release, minor release == patch release.

+1 to go to feature and patch release.

* 2.11.0 is a feature release;
* 3.0.0 is the first LTS release;
* 3.1.0 is a feature release;
* 3.2.0 is a feature release;
* 3.2.1 is a patch release;
* 4.0.0 is a LTS release.

## Compatibility between releases

When upgrading an existing cluster, it is important to upgrade components linearly.

> Before 3.0, upgrade should be done linearly through each minor version. For example, when upgrading from 2.8 to 2.10, it is important to upgrade to 2.9 before going to 2.10.

Reviewer 3: `Before 3.0, upgrade should be done linearly through each minor version.` - ditto
